File: 2523-closest-prime-numbers-in-range/2523-closest-prime-numbers-in-range.py
class Solution:
    def closestPrimes(self, left: int, right: int) -> List[int]:
        is_prime = [True] * (right + 1)
        if right >= 0: is_prime[0] = False
        if right >= 1: is_prime[1] = False
        
        for i in range(2, int(right**0.5) + 1):
            if is_prime[i]:
                for j in range(i * i, right + 1, i):
                    is_prime[j] = False
                    
        primes = [i for i in range(left, right + 1) if is_prime[i]]
        
        if len(primes) < 2:
            return [-1, -1]
            
        min_gap = float('inf')
        ans = [-1, -1]
        
        for i in range(1, len(primes)):
            gap = primes[i] - primes[i - 1]
            if gap < min_gap:
                min_gap = gap
                ans = [primes[i - 1], primes[i]]
            if min_gap <= 2:
                break
                
        return ans

File: 2523-closest-prime-numbers-in-range/2523-closest-prime-numbers-in-range.cpp
class Solution {
public:
    vector<int> closestPrimes(int left, int right) {
        vector<bool> isPrime(right + 1, true);
        if (right >= 0) isPrime[0] = false;
        if (right >= 1) isPrime[1] = false;
        
        for (int i = 2; i * i <= right; ++i) {
            if (isPrime[i]) {
                for (int j = i * i; j <= right; j += i) {
                    isPrime[j] = false;
                }
            }
        }
        
        
        vector<int> primes;
        for (int i = left; i <= right; ++i) {
            if (isPrime[i]) {
                primes.push_back(i);
            }
        }
        
        
        if (primes.size() < 2) {
            return {-1, -1};
        }
        
        int minGap = INT_MAX;
        vector<int> ans = {-1, -1};
        
        for (size_t i = 1; i < primes.size(); ++i) {
            int gap = primes[i] - primes[i - 1];
            if (gap < minGap) {
                minGap = gap;
                ans = {primes[i - 1], primes[i]};
            }
            
            if (minGap == 2 || minGap == 1) {
                break;
            }
        }
        
        return ans;
    }
};